### DISCUSSION TITLE
Local LSP server settings (tsserver preferences) + project-local override

### DISCUSSION THREAD

**Discussion by MatejBransky** [LazyVim/LazyVim#2703] [Q&A] [answered]
Hi, how can I add/change tsserver preferences?

From the mason I can see the path to the value in the server settings.

```lua
        ↓ typescript.preferences.importModuleSpecifier                                   default: "shortest"
          %typescript.preferences.importModuleSpecifier%
          type            string             
          possible values "shortest"          
                          "relative"          
                          "non-relative"      
                          "project-relative"  
```

But I don't know where to put it.

I've tried these and none of them worked. :(

```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        tsserver = {
          typescript = {
            preferences = {
              importModuleSpecifier = "non-relative",
            },
          },
        },
      },
    },
```
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        tsserver = {
          init_options = {
            typescript = {
              preferences = {
                importModuleSpecifier = "non-relative",
              },
            },
          },
        },
      },
    },
```
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        tsserver = {
          settings = {
            typescript = {
              ---@type _.lspconfig.settings.tsserver.Preferences
              preferences = {
                importModuleSpecifier = "non-relative",
              },
            },
          },
        },
      },
    },
```

Then I've tried to move the `tsserver` part under the `opts.servers`. No change. I'm desperate. There are too many possibilities where to put it.

Btw is it possible to reuse project-specific .vscode/settings.json file for tsserver settings?

**Comment by dpetka2001** [comment]
Searching for `importModuleSpecifier` gives back [this](https://github.com/LazyVim/LazyVim/discussions/1124). Try it out.

**Reply by MatejBransky** [reply]
The answer doesn't work either. The used value is still the "shortest". :( 

This is really difficult. Is there any guide how to configure typescript preferences because I couldn't find any working solution on github or google. 

The typescript-tools or vtsls are both well described and these preferences are easy to make but with the default tsserver in lspconfig I'm not able to find out how to do it.

**Reply by MatejBransky** [reply]
Maybe the suggested answer requires to configure it in some specific part of the lifecycle? I've tried it in the 'neovim/nvim-lspconfig''s `config = function (opts) ... end` then I've tried to directly set it under the `opts` key of the package. No difference, still doesn't work.

**Reply by dpetka2001** [reply]
I don't use Typescript so I don't know, unfortunately. I just did a quick search through the repo.

**Reply by MatejBransky** [reply]
And another thing - coc.nvim just works with the .vscode/settings.json which is really useful if you want to be productive ASAP without any knowledge of the neovim/lsp-config. Is something similar possible with the native LSP setup?

**Reply by MatejBransky** [reply]
I see, thanks for your help anyway. 🙂

**Reply by dpetka2001** [reply]
I've never used Vscode and its settings files. If you find coc.nvim easier for you to get work done, just use that. Just use the tools that help you get through your responsibilities and try new ones in your free time if you want to get familiar.

**Reply by MatejBransky** [reply]
I've tried coc but there were a lot of obstacles with non-working tools from the native LSP ecosystem and there were some bugs with keymaps so it was about the comparison what's more suitable for my work. Either option has some pros/cons but the overall evaluation looks better for the native LSP. The one issue which is not a blocker is to have project-specific settings for the LSP server and how to configure it based on the server's docs (original question of this discussion).

**Reply by dpetka2001** [reply]
You can also see this [issue](https://github.com/LazyVim/LazyVim/discussions/2043). That OP seemed to have found a working solution for him.

**Reply by MatejBransky** [reply]
Mmm. Doesn't work either.

**Reply by MatejBransky** [reply]
Mason still shows "shortest" and the code completion still suggests relative import.

**Reply by dpetka2001** [reply]
Mason shows the default value of the schema. Since the default option is `shortest` and you get relative import you must have changed something. Please note that the latest suggested answer from the other issue had configured the option you want to `relative`, because that's what that OP wanted. You did change the value of `importModuleSpecifierPreference`, right?

**Reply by MatejBransky** [reply]
🤦‍♂️ ..ok, I've missed the opposite value in the last suggested answer. Yep, that was it. It works! 🎉 Thank you very much! 🙂

**Reply by dpetka2001** [reply]
@MatejBransky Also, if you'd like some project-specific settings, you should look into this [PR](https://github.com/LazyVim/LazyVim/pull/2115), which provides such a functionality.

**Reply by MatejBransky** [reply, marked as answer]
Cool! I didn't know about it! So the final solution I'm using: 

I have in a project the `.lazy.lua` file which contains the working solution you sent me. 

`<project-dir>/.lazy.lua`
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        tsserver = {
          init_options = {
            preferences = {
              importModuleSpecifierPreference = "non-relative",
              importModuleSpecifierEnding = "minimal",
            },
          },
        },
      },
    },
  },
}
```

And this works only if you enable/add LazyExtra plugin in the `lazy.nvim` file in imports and it has to be done like that so you can explicitly define that it should be imported as the last plugin.

`~/.config/nvim/lua/config/lazy.lua`
```lua
require("lazy").setup({
  spec = {
    -- add LazyVim and import its plugins
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    -- import any extras modules here
    { import = "lazyvim.plugins.extras.lang.typescript" },
    { import = "lazyvim.plugins.extras.linting.eslint" },
    { import = "lazyvim.plugins.extras.formatting.prettier" },
    { import = "lazyvim.plugins.extras.lang.json" },
    -- { import = "lazyvim.plugins.extras.lang.go" },
    -- { import = "lazyvim.plugins.extras.coding.copilot" },
    { import = "lazyvim.plugins.extras.ui.mini-animate" },
    -- import/override with your plugins
    { import = "plugins" },
    -- this allows to override nvim/plugins configs locally with the `<project-dir>/.lazy.lua` file
    { import = "lazyvim.plugins.extras.lazyrc" },
  },
  -- ...
})
```

Thank you for your time! 🙏

**Reply by dpetka2001** [reply]
Maybe you should select your own answer as the suggested one, since I just redirected you to another post. This way, another user that may have the same issue can directly see the answer here, instead of having to visit another link if my answer is the suggested one. Because there might be cases, where someone might just read the suggested answer and skip the rest. I think it makes more sense this way.
